core/src/main/scala/org/apache/spark/h2o/backends/internal/InternalReadConverterContext.scala

```diff
@@ -37,39 +38,106 @@ class InternalReadConverterContext(override val keyName: String, override val ch
 
   override def isNA(columnNum: Int): Boolean = chks(columnNum).isNA(rowIdx)
 
-  private def get[T](columnNum: Int, read: Chunk => T): Option[T] =
-    if (isNA(columnNum)) {
-      None
-    } else {
-      Option(read(chks(columnNum)))
-    }
-
-  override def getLong(columnNum: Int): Option[Long] =  get(columnNum, _.at8(rowIdx))
-  override def getDouble(columnNum: Int): Option[Double] = get(columnNum, _.atd(rowIdx))
+  private def getOption[T](read: Chunk => T)(columnNum: Int): Option[T] = {
+    for {
+      chunk <- Option(chks(columnNum)) if !chunk.isNA(rowIdx)
+      data <- Option(read(chunk))
+    } yield data
+  }
 
-  // TODO(vlad): try to move out all this logic to prepared extractors
-  override def getString(columnNum: Int): Option[String] = get(columnNum, chunk =>
-    {
+  private def getSimple[T](defaultValue: T, read: Chunk => T)(columnNum: Int): T = {
     val chunk = chks(columnNum)
-    val vector = chunk.vec()
-    if (vector.isCategorical) {
-      val str = vector.domain()(chunk.at8(rowIdx).toInt)
-      str
-    } else if (vector.isString) {
-      val valStr = new BufferedString()
-      chunk.atStr(valStr, rowIdx) // TODO improve this.
-      valStr.toString
-    } else if (vector.isUUID) {
-      val uuid = new UUID(chunk.at16h(rowIdx), chunk.at16l(rowIdx))
-      uuid.toString
-    } else {
-      assert(assertion = false, s"Should never be here, type is ${vector.get_type_str()}")
-      // TODO(vlad): this is temporarily here, to provide None is string is missing
-      null
-    }
-    })
+    if (chunk.isNA(rowIdx)) defaultValue else read(chunk)
+  }
+
+  override def getLong(columnNum: Int): Option[Long] = {
```

Comment:
I would write it just in one like, just to be consistent with the rest of the code 

`override def getLong(columnNum: Int): Option[Long] = getOption(_.at8(rowIdx))(columnNum)`